### PR TITLE
Gather allocated CPUs from cgroups

### PR DESCRIPTION
**Pull request by swatisehgal**
Signed-off-by: Swati Sehgal <swsehgal@redhat.com>

### CODE REVIEW

File: pkg/finder/finder.go [hunk 1]
```diff
@@ -26,9 +28,10 @@ import (
 )
 
 const (
-	defaultTimeout   = 5 * time.Second
-	ns               = "resource-topology-exporter"
-	deviceNamePrefix = "EXAMPLECOMDEVICE"
+	defaultTimeout     = 5 * time.Second
+	ns                 = "resource-topology-exporter"
+	cgroupCpuSetPrefix = "/host-sys/fs/cgroup/cpuset/"
```
Comment (ffromani) [inline pkg/finder/finder.go]: I'd not use an absolute path here, because doing so leaks in the pod definition. In other word, if the host sysfs is mounted somewhere else than on `/host-sys`, this code unnecessarily breaks. We should get the sysfs mount point as parameter

File: pkg/finder/finder.go [hunk 2]
```diff
@@ -464,3 +472,23 @@ func getContainerEnvironmentVariables(ci ContainerInfo) map[string]string {
 	// nothing else to try, give up and fail!
 	return nil
 }
+
+func getAllocatedCPUs(cgroupsPath string, CRIEndpointPath string) (string, error) {
+	var cgroupAbsolutePath string
+	switch CRIEndpointPath {
+	case "/host-run/crio/crio.sock":
+		cgroupPathFromCRI := strings.Replace(cgroupsPath, "slice:crio:", "slice/crio-", 1)
+		cgroupAbsolutePath = fmt.Sprint(cgroupCpuSetPrefix, "kubepods.slice/", cgroupPathFromCRI, ".scope", CPUsetSuffix)
```
Comment (ffromani) [inline pkg/finder/finder.go]: I'd use `filepath.Join`

File: pkg/finder/finder.go [hunk 3]
```diff
@@ -464,3 +472,23 @@ func getContainerEnvironmentVariables(ci ContainerInfo) map[string]string {
...
+
+	case "/host-run/containerd/containerd.sock":
+		cgroupAbsolutePath = fmt.Sprint(cgroupCpuSetPrefix, cgroupsPath, CPUsetSuffix)
```
Comment (ffromani) [inline pkg/finder/finder.go]: ditto

File: pkg/finder/finder.go [hunk 4]
```diff
@@ -464,3 +472,23 @@ func getContainerEnvironmentVariables(ci ContainerInfo) map[string]string {
 	// nothing else to try, give up and fail!
 	return nil
 }
+
+func getAllocatedCPUs(cgroupsPath string, CRIEndpointPath string) (string, error) {
```
Comment (ffromani) [inline pkg/finder/finder.go]: this function does two main things:
1. it finds the full cgroup path given 
a. the host sysfs mount path
b. the detected container runtime
2. reads the cpu allocation from the path learned from the step above

Now, turns out that everything is actually generic provided we translate the path we learn from the container info into a real path. Right now we learn about runtime guessing from the endpointpath. But nothing in the actual path except for the conventions guarantees that we can safely deduct the runtime from the path. In other words, nothing prevents someone to configure the kubelet to use something like `/run/cri/runtime.sock` independenty from crio or containerd.

TL;DR: I think we need an explicit flag, that's the only way we can really be sure.

File: pkg/finder/finder.go [hunk 5]
```diff
@@ -464,3 +472,23 @@ func getContainerEnvironmentVariables(ci ContainerInfo) map[string]string {
...
+	var cgroupAbsolutePath string
+	switch CRIEndpointPath {
+	case "/host-run/crio/crio.sock":
+		cgroupPathFromCRI := strings.Replace(cgroupsPath, "slice:crio:", "slice/crio-", 1)
```
Comment (ffromani) [inline pkg/finder/finder.go]: once we have a flag we can extract the path extraction from this function and set it explicitely in main, so the runtime-specific step is clearly separated

**Review by ffromani — COMMENTED** [review]
The commit message could be a little less terse :)
In general I believe this patch works, but with some changes we can make it more future proof.
The core idea is to explicitely distinguish between runtimes because, despite all the best efforts, we will have (small) implementation differences between runtimes. Sometimes we can autodetect them and handle transparently; sometimes we can't.

File: pkg/finder/finder.go (hunk 5 above)
Comment (ffromani) [reply pkg/finder/finder.go]: This turned out very poorly written, it's probably simpler to demonstrate the concept with this (completely untested): https://gist.github.com/fromanirh/5de2ef51be8d6f3a978b8d5bf3418b5d

File: pkg/finder/finder.go (hunk 1 above)
Comment (swatisehgal) [reply pkg/finder/finder.go]: Ah yeah, that makes sense, will update this!

File: pkg/finder/finder.go (hunk 2 above)
Comment (swatisehgal) [reply pkg/finder/finder.go]: Will do

File: pkg/finder/finder.go (hunk 5 above)
Comment (swatisehgal) [reply pkg/finder/finder.go]: So elegant! Will update the code, Thanks for the input!

File: pkg/finder/finder.go (hunk 4 above)
Comment (swatisehgal) [reply pkg/finder/finder.go]: I understand, I wasn't a huge fan of using the socket names either, will add a flag for explicitly specifying the container runtime

**Comment by swatisehgal** [comment]
@fromanirh PTAL and let me know if it looks better now. I have tested these changes on a cluster and it works as expected :)

File: pkg/finder/finder.go [hunk 6]
```diff
@@ -114,6 +119,14 @@ func NewFinder(args Args, pciResMapConf map[string]string) (CRIFinder, error) {
 	}
 	var err error
 
+ 	// At this stage, we only support containerd and cri-o
+	if args.ContainerRuntime == "containerd" {
```
Comment (swatisehgal) [inline pkg/finder/finder.go:123]: @fromanirh Let me know if this check is reasonable to do or if you think we should do it in another way.

**Comment by ffromani** [comment]
It does indeed look much better, thanks for the updates. I've added inline comments on the new commit, there are a few more low-hanging fruits improvements for a better UX.

File: main.go [hunk 7]
```diff
@@ -101,6 +104,7 @@ func argsParse(argv []string) (finder.Args, error) {
 		args.SRIOVConfigFile = path
 	}
 	args.SysfsRoot = arguments["--sysfs"].(string)
+	args.ContainerRuntime = arguments["--container-runtime"].(string)
```
Comment (ffromani) [inline main.go]: nice, but let's add an explicit validation here, for better UX and more robust code. So the user can know asap which runtimes are supported and how we identify them.

File: pkg/finder/finder.go (hunk 6 above)
Comment (ffromani) [reply pkg/finder/finder.go:123]: it is totally reasonnable, let's just use package-level constants and let's add explicit validation instead of silently handling errors. E.g. if the user provides `--container-runtime=foobar` we should fail immediately and not assume it means "cri-o".

**Review by ffromani — APPROVED** [review]
Very nice! Feel free to squash commits as you see most fit!

File: main.go (hunk 7 above)
Comment (swatisehgal) [reply main.go]: I have added another commit for input validation, hope it looks okay now.

File: main.go [hunk 8]
```diff
@@ -104,7 +104,11 @@ func argsParse(argv []string) (finder.Args, error) {
 		args.SRIOVConfigFile = path
 	}
 	args.SysfsRoot = arguments["--sysfs"].(string)
-	args.ContainerRuntime = arguments["--container-runtime"].(string)
+	runtime := arguments["--container-runtime"].(string)
+	if !(runtime == "containerd" || runtime == "cri-o") {
```
Comment (swatisehgal) [inline main.go:108]: Its equivalent to : `if runtime != "containerd" && runtime != "cri-o" {`
Can do this too, not really sure which is conventionally better

File: pkg/finder/finder.go (hunk 6 above)
Comment (swatisehgal) [reply pkg/finder/finder.go:123]: Updated now, Thanks for your feedback.

File: main.go (hunk 8 above)
Comment (ffromani) [reply main.go:108]: both forms are fine. The one you have in the comments is probably more widespread and thus a bit easier to read, but the other is not wrong. I'd jyst move forward and revisit when we add the (much needed) tests.

Comment (swatisehgal) [reply main.go:108]: Okay, We can update later when needed.
